File: examples/checks/superdense_coding.py
# Independent Python checks for the superdense_coding example.
from __future__ import annotations

import re
from collections import Counter

from .common import run_checks


Pair = tuple[bool, bool]


def relation(rows: list[list[bool]]) -> set[Pair]:
    # The JSON serializes each fact row as two mobit pairs: [a,b,c,d]
    # means relation contains (a,b) and (c,d). Duplicates collapse.
    out: set[Pair] = set()
    for row in rows:
        if len(row) != 4:
            raise ValueError("relation rows must contain four booleans")
        out.add((row[0], row[1]))
        out.add((row[2], row[3]))
    return out


def compose(first: set[Pair], second: set[Pair]) -> set[Pair]:
    return {(a, d) for (a, b) in first for (c, d) in second if b == c}


def derive(data: dict):
    states = {name: relation(rows) for name, rows in data["States"].items()}
    primitive = {name: relation(rows) for name, rows in data["Primitive"].items()}
    composed = {
        "kg": compose(primitive["g"], primitive["k"]),
        "gk": compose(primitive["k"], primitive["g"]),
    }
    lookup = lambda name: primitive.get(name, composed.get(name))
    alice = {int(k): v for k, v in data["AliceOps"].items()}
    # The Go translation names the four Bob tests by the relations that accept
    # each encoded support after the joint measurement.
    bob = {0: "gk", 1: "k", 2: "g", 3: "id"}
    counts = Counter()
    candidates = []
    encoded = {}
    for msg in range(4):
        alice_rel = lookup(alice[msg])
        encoded[msg] = set()
        for shared_a, shared_b in sorted(states["R"]):
            for move_a, move_b in sorted(alice_rel):
                if move_a != shared_a:
                    continue
                encoded[msg].add((move_b, shared_b))
                for decoded in range(4):
                    if (move_b, shared_b) in lookup(bob[decoded]):
                        counts[(msg, decoded)] += 1
                        candidates.append((msg, decoded, shared_a, move_b, shared_b))
    survivors = {msg: decoded for msg in range(4) for decoded in range(4) if counts[(msg, decoded)] % 2 == 1}
    return states, primitive, composed, counts, candidates, survivors, encoded


def parse_survivors(answer: str):
    return {int(a): int(b) for a, b in re.findall(r"^\s*(\d+) dqc:superdense-coding (\d+)", answer, flags=re.MULTILINE)}


def run(ctx):
    data = ctx.load_input()
    states, primitive, composed, counts, candidates, survivors, encoded = derive(data)
    reported = parse_survivors(ctx.answer)
    diagonal_counts = [counts[(i, i)] for i in range(4)]
    offdiag_even = all(counts[(i, j)] % 2 == 0 for i in range(4) for j in range(4) if i != j)

    checks = [
        ("shared entanglement R contains exactly |0,0) and |1,1)", states["R"] == {(False, False), (True, True)}),
        ("composition KG is recomputed by composing G then K", composed["kg"] == {(False, False), (False, True), (True, False)}),
        ("composition GK is recomputed by composing K then G", composed["gk"] == {(False, True), (True, False), (True, True)}),
        ("the raw superdense rule creates 24 candidate derivations", len(candidates) == 24),
        ("GF(2) cancellation leaves odd diagonal and even off-diagonal counts", diagonal_counts == [1, 1, 1, 1] and offdiag_even),
        ("reported surviving decoded messages match the parity survivors", reported == survivors == {0: 0, 1: 1, 2: 2, 3: 3}),
        ("the four Alice operations produce distinct encoded supports", len({tuple(sorted(v)) for v in encoded.values()}) == 4),
        ("the JSON relation facts match the primitive teaching-model relations", primitive["id"] == states["R"] and primitive["g"] == states["S"] and primitive["k"] == states["U"]),
    ]
    return run_checks(checks)

File: examples/checks/superdense_coding.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

func checkSuperdense(ctx *Context) []Check {
	d := ctx.M()
	states := map[string]map[mobPair]bool{}
	for name, rows := range asMap(d["States"]) {
		states[name] = rel(asSlice(rows))
	}
	prim := map[string]map[mobPair]bool{}
	for name, rows := range asMap(d["Primitive"]) {
		prim[name] = rel(asSlice(rows))
	}
	comp := map[string]map[mobPair]bool{"kg": composeRel(prim["g"], prim["k"]), "gk": composeRel(prim["k"], prim["g"])}
	lookup := func(name string) map[mobPair]bool {
		if r, ok := prim[name]; ok {
			return r
		}
		return comp[name]
	}
	alice := map[int]string{}
	for k, v := range asMap(d["AliceOps"]) {
		alice[parseInt(k)] = str(v)
	}
	bob := map[int]string{0: "gk", 1: "k", 2: "g", 3: "id"}
	counts := map[[2]int]int{}
	candidates := 0
	encoded := map[int]map[mobPair]bool{}
	for msg := 0; msg < 4; msg++ {
		encoded[msg] = map[mobPair]bool{}
		for sh := range states["R"] {
			for mv := range lookup(alice[msg]) {
				if mv.a != sh.a {
					continue
				}
				encoded[msg][mobPair{mv.b, sh.b}] = true
				for dec := 0; dec < 4; dec++ {
					if lookup(bob[dec])[mobPair{mv.b, sh.b}] {
						counts[[2]int{msg, dec}]++
						candidates++
					}
				}
			}
		}
	}
	survivors := map[int]int{}
	diag := []int{}
	offEven := true
	for i := 0; i < 4; i++ {
		diag = append(diag, counts[[2]int{i, i}])
		for j := 0; j < 4; j++ {
			if i != j && counts[[2]int{i, j}]%2 != 0 {
				offEven = false
			}
			if counts[[2]int{i, j}]%2 == 1 {
				survivors[i] = j
			}
		}
	}
	reported := map[int]int{}
	for _, m := range reAll(ctx.Answer, `(?m)^\s*(\d+) dqc:superdense-coding (\d+)`) {
		reported[parseInt(m[1])] = parseInt(m[2])
	}
	encodedSupports := map[string]bool{}
	for _, v := range encoded {
		parts := []string{}
		for p := range v {
			parts = append(parts, fmt.Sprintf("%v/%v", p.a, p.b))
		}
		sort.Strings(parts)
		encodedSupports[strings.Join(parts, ",")] = true
	}
	return []Check{{"shared entanglement R contains exactly |0,0) and |1,1)", relEq(states["R"], map[mobPair]bool{{false, false}: true, {true, true}: true})}, {"composition KG is recomputed by composing G then K", relEq(comp["kg"], map[mobPair]bool{{false, false}: true, {false, true}: true, {true, false}: true})}, {"composition GK is recomputed by composing K then G", relEq(comp["gk"], map[mobPair]bool{{false, true}: true, {true, false}: true, {true, true}: true})}, {"the raw superdense rule creates 24 candidate derivations", candidates == 24}, {"GF(2) cancellation leaves odd diagonal and even off-diagonal counts", intsEq(diag, []int{1, 1, 1, 1}) && offEven}, {"reported surviving decoded messages match the parity survivors", mapIntEq(reported, survivors) && mapIntEq(survivors, map[int]int{0: 0, 1: 1, 2: 2, 3: 3})}, {"the four Alice operations produce distinct encoded supports", len(encodedSupports) == 4}, {"the JSON relation facts match the primitive teaching-model relations", relEq(prim["id"], states["R"]) && relEq(prim["g"], states["S"]) && relEq(prim["k"], states["U"])}}
}
